### PR TITLE
Feature/118 smb multi thread

### PR DESCRIPTION
**Pull request by stgmsa**
# Description
introduce the multi machine account support for NTLM Auth API

# Impacts
* The ntlm auth api service management - gunicorn introduced
* The ntlm auth api workflow - multiple worker process will be launched
* The ntlm auth api configuration - "additional machine accounts" added.

# NEW Package(s) required
python gunicorn
python redis

# Delete branch after merge
YES

# Checklist
- [yes] Document the feature
- [no] Add OpenAPI specification
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

## Enhancements
parallel ntlm auth API supported

### CODE REVIEW

File: addons/upgrade/to-14.1-adds-domain-multi-machine-account-support.pl [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+=cut
+
+use strict;
+use warnings;
+use lib qw(/usr/local/pf/lib /usr/local/pf/lib_perl/lib/perl5);
+use pf::IniFiles;
+use pf::file_paths qw($domain_config_file);
+use pf::util;
+use Sys::Hostname;
```
Comment (fdurand) [inline addons/upgrade/to-14.1-adds-domain-multi-machine-account-support.pl]: Not needed

File: bin/pyntlm_auth/gunicorn.conf.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+workers = worker_num
+worker_class = 'sync'  # use sync, do not use 'gevent', or 'eventlet' due to block operations.
+timeout = 30
+graceful_timeout = 10
+
+accesslog = '-'  # to stdout
+errorlog = '-'  # to stdout/err
+loglevel = 'info'  # debug info warning error critical
+capture_output = False  # do not forward logs from stdout / err to log files.
+# access_log_format = '%(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"'     # default access log format.
```
Comment (fdurand) [inline bin/pyntlm_auth/gunicorn.conf.py]: remove if not needed

File: bin/pyntlm_auth/gunicorn.conf.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+access_log_format = '%(h)s %(l)s %(u)s %(p)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"'
+
+keepalive = 2
+max_requests = 10000
+max_requests_jitter = 50
+
+daemon = False
+# pidfile = '/tmp/gunicorn.pid'
```
Comment (fdurand) [inline bin/pyntlm_auth/gunicorn.conf.py]: remove if not needed

File: bin/pyntlm_auth/redis_client.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import redis
+
+r = redis.StrictRedis(
+    host='100.64.0.1', port=6379, db=0,
```
Comment (fdurand) [inline bin/pyntlm_auth/redis_client.py]: we should use a env variable for that.

File: html/pfappserver/lib/pfappserver/Form/Config/Domain.pm [hunk 5]
```diff
@@ -258,6 +258,14 @@ has_field 'machine_account_password' =>
         tags => { after_element => \&help,
             help => 'Password of the machine account to be added to Active Directory.' },
     );
+has_field 'additional_machine_accounts' =>
+    (
+        type => 'PosInteger',
+        label => 'Additional machine account for NTLM authentication',
+        default => 0,
+        tags => { after_element => \&help,
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Config/Domain.pm:266]: Not necessary anymore since the help text message is in the frontend code now.

**Review by fdurand — CHANGES_REQUESTED** [review]

**Comment by E-Thang** [comment]
I'm unable to start the ntlm-auth-api on 14.1 after this was merged. The file config_loader.py is definitely there on the host. Not sure why it's saying it isn't available to the container. I can't say what actually is in the container though. I tried adding to the Dockerfile to copy everything from the  bin/pyntlm_auth/ during build. Then rebuilt the container and deleted the old images. Didn't change a thing, I don't know much about docker though.
```
2024-11-08T23:19:47.545270-08:00 boomer ntlm-auth-api-domain[675753]: Running with args --sig-proxy=true --rm --name=ntlm-auth-api-OITAD --add-host=containers-gateway.internal:host-gateway -h boomer  -v /var/lib/mysql:/var/lib/mysql -v /etc/sudoers:/etc/sudoers -v /etc/sudoers.d/:/etc/sudoers.d/ -v /usr/local/fingerbank/conf:/usr/local/fingerbank/conf -v /usr/local/fingerbank/db:/usr/local/fingerbank/db -v /usr/local/pf/var/run:/usr/local/pf/var/run -ePF_UID=996 -e PF_GID=995 -eFINGERBANK_UID=999 -e FINGERBANK_GID=996 -eIS_A_CLASSIC_PF_CONTAINER=yes -eTZ=America/Los_Angeles -v /usr/local/pf/conf:/usr/local/pf/conf -v /usr/local/pf/logs:/usr/local/pf/logs -v /usr/local/pf/var/conf:/usr/local/pf/var/conf -v /var/log:/var/log:ro -p 5000:5000 -e LISTEN=5000 -e IDENTIFIER=OITAD
2024-11-08T23:19:47.568669-08:00 boomer ntlm-auth-api-domain[675764]: Error response from daemon: No such container: ntlm-auth-api-OITAD
2024-11-08T23:19:47.568669-08:00 boomer ntlm-auth-api-domain[675764]: Error response from daemon: No such container: 1
2024-11-08T23:19:48.290363-08:00 boomer ntlm-auth-api-domain[675773]: Failed to read config file: /usr/local/pf/bin/pyntlm_auth/gunicorn.conf.py
2024-11-08T23:19:48.291090-08:00 boomer ntlm-auth-api-domain[675773]: Traceback (most recent call last):
2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:  File "/usr/lib/python3/dist-packages/gunicorn/app/base.py", line 111, in get_config_from_filename
2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:    spec.loader.exec_module(mod)
2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:  File "/usr/local/pf/bin/pyntlm_auth/gunicorn.conf.py", line 6, in <module>
2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:    import config_loader
2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]: ModuleNotFoundError: No module named 'config_loader'
```

**Comment by stgmsa** [comment]
> I'm unable to start the ntlm-auth-api on 14.1 after this was merged. The file config_loader.py is definitely there on the host. Not sure why it's saying it isn't available to the container. I can't say what actually is in the container though. I tried adding to the Dockerfile to copy everything from the bin/pyntlm_auth/ during build. Then rebuilt the container and deleted the old images. Didn't change a thing, I don't know much about docker though.
> 
> ```
> 2024-11-08T23:19:47.545270-08:00 boomer ntlm-auth-api-domain[675753]: Running with args --sig-proxy=true --rm --name=ntlm-auth-api-OITAD --add-host=containers-gateway.internal:host-gateway -h boomer  -v /var/lib/mysql:/var/lib/mysql -v /etc/sudoers:/etc/sudoers -v /etc/sudoers.d/:/etc/sudoers.d/ -v /usr/local/fingerbank/conf:/usr/local/fingerbank/conf -v /usr/local/fingerbank/db:/usr/local/fingerbank/db -v /usr/local/pf/var/run:/usr/local/pf/var/run -ePF_UID=996 -e PF_GID=995 -eFINGERBANK_UID=999 -e FINGERBANK_GID=996 -eIS_A_CLASSIC_PF_CONTAINER=yes -eTZ=America/Los_Angeles -v /usr/local/pf/conf:/usr/local/pf/conf -v /usr/local/pf/logs:/usr/local/pf/logs -v /usr/local/pf/var/conf:/usr/local/pf/var/conf -v /var/log:/var/log:ro -p 5000:5000 -e LISTEN=5000 -e IDENTIFIER=OITAD
> 2024-11-08T23:19:47.568669-08:00 boomer ntlm-auth-api-domain[675764]: Error response from daemon: No such container: ntlm-auth-api-OITAD
> 2024-11-08T23:19:47.568669-08:00 boomer ntlm-auth-api-domain[675764]: Error response from daemon: No such container: 1
> 2024-11-08T23:19:48.290363-08:00 boomer ntlm-auth-api-domain[675773]: Failed to read config file: /usr/local/pf/bin/pyntlm_auth/gunicorn.conf.py
> 2024-11-08T23:19:48.291090-08:00 boomer ntlm-auth-api-domain[675773]: Traceback (most recent call last):
> 2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:  File "/usr/lib/python3/dist-packages/gunicorn/app/base.py", line 111, in get_config_from_filename
> 2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:    spec.loader.exec_module(mod)
> 2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
> 2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
> 2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:  File "/usr/local/pf/bin/pyntlm_auth/gunicorn.conf.py", line 6, in <module>
> 2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]:    import config_loader
> 2024-11-08T23:19:48.291200-08:00 boomer ntlm-auth-api-domain[675773]: ModuleNotFoundError: No module named 'config_loader'
> ```

Hi @E-ThanG 
Thanks for the feedback, seems to be a dependency / FS issue. We'll investigate and will reach you back soon.

**Comment by E-Thang** [comment]
I was able to get it to start by adding ` sys.path.append("/usr/local/pf/bin/pyntlm_auth") ` to gunicorn.conf.py just before it imports config_loader. It also needed python3-redis added to the Dockerfile.

Now that it starts, it still won't work. It's complaining that  "[WARNING] failed to bind machine account: no available accounts, retrying."

I tried with 0, 1, 2, and 8 accounts. It creates them in AD but can't seem to use them. It's trying with the account  hostname$ and with hostname-0$.

```
2024-11-10T19:53:58.340448-08:00 boomer ntlm-auth-api-domain[74418]: [2024-11-10 19:53:58 -0800] [8] [INFO] deal machine account test for: BOOMER$ with password 'xxx'
2024-11-10T19:53:58.460697-08:00 boomer ntlm-auth-api-domain[74418]: [2024-11-10 19:53:58 -0800] [9] [INFO] deal machine account test for: BOOMER-0$ with password 'xxx'

```

**Pull request merged**